compiler/src/vm/heap/object.rs

```diff
@@ -0,0 +1,238 @@
+use super::{pointer::Pointer, Heap};
+use crate::{
+    builtin_functions::BuiltinFunction,
+    compiler::{
+        hir_to_lir::HirToLir,
+        lir::{Instruction, Lir},
+    },
+    database::Database,
+    module::Module,
+};
+use itertools::Itertools;
+use num_bigint::BigInt;
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+    ops::Deref,
+};
+
+#[derive(Clone)]
+pub struct Object {
+    pub reference_count: usize,
+    pub data: Data,
+}
+#[derive(Clone)]
+pub enum Data {
+    Int(Int),
+    Text(Text),
+    Symbol(Symbol),
+    Struct(Struct),
+    Closure(Closure),
+    Builtin(Builtin),
+}
+
+#[derive(Clone)]
+pub struct Int {
+    pub value: BigInt,
+}
+
+#[derive(Clone)]
+pub struct Text {
+    pub value: String,
+}
+
+#[derive(Clone)]
+pub struct Symbol {
+    // TODO: Choose a more efficient representation.
+    pub value: String,
+}
+
+#[derive(Default, Clone)]
+pub struct Struct {
+    // TODO: Choose a more efficient representation.
+    pub buckets: HashMap<u64, Vec<(Pointer, Pointer)>>, // hash -> list of keys and values
```

Comment:
I think I'd rather revert this to a single `Vec<u64, Pointer, Pointer>`. I'm assuming that our structs usually contain few keys so a `HashMap` is probably overkill, especially if we sort the entries. Also, this `HashMap` representation with `Vec`s adds 24 bytes and another level of indirection to each entry with distinct hashes.

With a single, sorted `Vec`, we could reduce the previous two linear passes to one binary search plus a linear pass within hash collisions by sorting the struct by hash and then the raw key's pointer:

```rust
        let hash = key.hash(heap);
        let first_index_for_hash = self.fields
            .binary_search_by_key(
                (&hash, Pointer::null()),
                |(hash, key, _)| (hash, key),
            )
            .into_ok_or_error();
            // needs `#![feature(result_into_ok_or_err)]`
            // https://github.com/rust-lang/rust/issues/82223
        let mut index = first_index_for_hash;
        for (existing_index, (existing_hash, existing_key, existing_value)) in &mut self.fields[first_index_for_hash..].iter_mut().enumerate() {
            if hash != *existing_hash {
                index = existing_index;
                break;
            }
            if existing_key.equals(heap, key) {
                *existing_value = value;
                return;
            }
            if existing_key.raw < key.raw {
                index = existing_index;
            }
        }
        self.fields.insert(existing_index, (hash, key, value));
```

Reply:
Hmm. Our values are not interned, so we can't simply search by pointers – if I'm looking for whatever is at key `123`, we want to consider that key equal to another `123` with a different address. So, we'd have to use `binary_search_by` and do the comparison ourselves. Except, we can't compare values. In particular, closures have no well-defined ordering and I don't think they should.

We could of course only binary search for the hash. Then, once we found a hash, we look in both directions of the vec (assuring the hash is still the same) and compare our value to the others that exist there. Do you think it's worth implementing this now or should we leave that for a later time? I can imagine it'll be a lot more code than the current 6 line function.

Reply:
Have a look at the code I posted above. It implements what you describe in your second paragraph: We do a binary search to find the first matching hash in the struct and then traverse linearly to find the matching key. The key ordering is only based on the pointer values, not the actual key values. 

It's not that much code, though I haven't tested it and the getter also has to be modified

Reply:
If there are multiple matches, then `binary_search_by_key` returns any one of them, not the first one. But I guess I can implement it this way

Reply:
Ah nevermind, I thought that was the code for retrieving an item

Reply:
Yes, that's why I'm searching for the hash in combination with a null pointer: That way, we always find the left-most element with a matching hash (if it exists)

Reply:
Ohhhh I didn't catch that. Makes sense (although I found an even more readable version)